src/express.re

```diff
@@ -1,39 +1,147 @@
 /** abstract type which ensure middleware function must either
-    call the [next] function or one of the [send] function on the 
-    response object. 
+    call the [next] function or one of the [send] function on the
+    response object.
 
-    This should be a great argument for OCaml, the type system 
-    prevents silly error which in this case would make the server hang */ 
+    This should be a great argument for OCaml, the type system
+    prevents silly error which in this case would make the server hang */
 
 /** TODO : maybe this should be a more common module like bs-error */
 module Error = {
 
   type t;
   /** Error type */
 
-  external message : t => option string = "" 
+  external message : t => option string = ""
     [@@bs.send][@@bs.return null_undefined_to_opt];
-  external name : t => option string = "" 
+  external name : t => option string = ""
     [@@bs.send][@@bs.return null_undefined_to_opt];
 };
 
 module Request = {
   type t;
-  
+
   type params = Js_dict.t Js_json.t;
 
   external params : t => params = "" [@@bs.get];
-  /** [params request] return the JSON object filled with the 
+  /** [params request] return the JSON object filled with the
       request parameters */
 
   external asJsonObject : t => Js_dict.t Js_json.t = "%identity";
-  /** [asJsonObject request] casts a [request] to a JSON object. It is 
-      common in Express application to use the Request object as a 
-      placeholder to maintain state through the various middleware which 
-      are executed. */ 
-
-  /** more property method coming later */
+  /** [asJsonObject request] casts a [request] to a JSON object. It is
+      common in Express application to use the Request object as a
+      placeholder to maintain state through the various middleware which
+      are executed. */
+
+  external baseUrl: t => string = "" [@@bs.get];
+  /** [baseUrl request] returns the 'baseUrl' property */
+
+  external hostname : t => string = "" [@@bs.get];
+  /** [hostname request] Contains the hostname derived from the Host
+      HTTP header.*/
+
+  external ip : t => string = "" [@@bs.get];
+  /** [ip request] Contains the remote IP address of the request.*/
+
+  external fresh : t => bool =  "" [@@bs.get];
+  /** [fresh request] returns [true] whether the request is "fresh" */
+
+  external stale : t => bool =  "" [@@bs.get];
+  /** [stale request] returns [true] whether the request is "stale"*/
+
+  external methodRaw : t => string = "method" [@@bs.get];
+
+  type method_ =
+    | Get
+    | Post
+    | Put
+    | Delete
+    | Head
+    | Options
+    | Trace
+    | Connect;
+
+  let method_ : t => method_ = fun req => {
+    switch (methodRaw req) {
+      | "GET" => Get;
+      | "POST" => Post;
+      | "PUT" => Put;
+      | "DELETE" => Delete;
+      | "HEAD" => Head;
+      | "OPTIONS" => Options;
+      | "TRACE" => Trace;
+      | "CONNECT" => Connect;
+      | s => failwith ("Express.Request.method_ Unexpected method: " ^ s);
+    }};
```

Comment:
I've been wondering it might be a better idea to model string and number enums as constants instead of a variant. It means giving up exhaustiveness checking, and might not seem to make much sense in this specific case, but as a general FFI convention it would significantly improve the amount of generated code in other cases (see bs-webapi), and would be more consistent with bitflag constants. And if the constant types are exposed as aliases of the underlying type it would remove the need for the `raw` functions here.

I'm not saying you should change this, just thinking out loud, and perhaps raising it for discussion in the wrong place entirely.